Contributing branches and reverting (#9577)
NixOS/nix: doc/manual/src/contributing/hacking.md

Releases have a precomputed `rl-MAJOR.MINOR.md`, and no `rl-next.md`.
Set `buildUnreleasedNotes = true;` in `flake.nix` to build the release notes on the fly.

## Branches

> - `master`

Reviewer: Maybe link the branch names to GitHub? There are even search queries you can link.

  The main development branch. All changes are approved and merged here.
  When developing a change, create a branch based on the latest `master`.

>   It is [kept](#reverting) in a release-worthy state.

Reviewer: ```suggestion
  Maintainers try to [keep it in a release-worthy state](#reverting).
```

Use link texts that can stand on their own.

- `maintenance-*.*`

  These branches are the subject of backports only, and are
  also [kept](#reverting) in a release-worthy state.

  See [`maintainers/backporting.md`](https://github.com/NixOS/nix/blob/master/maintainers/backporting.md)

- `latest-release`

  The latest patch release of the latest minor version.

  See [`maintainers/release-process.md`](https://github.com/NixOS/nix/blob/master/maintainers/release-process.md)

- `backport-*-to-*`

  Generally branches created by the backport action.

  See [`maintainers/backporting.md`](https://github.com/NixOS/nix/blob/master/maintainers/backporting.md)

- _other_

>   Branches that do not conform to the above patterns should be feature branches.

Reviewer: IMHO those should not exist at all, and feature branches should come from forks. Since this documentation is descriptive rather than prescriptive, I won't block that with my desire to discuss this, but I'd still like to discuss it eventually.

Author: Forks in organizations are problematic because they can't be configured to allow contributions from upstream maintainers. We might suggest using personal forks.

Reviewer 2: I also tend to use feature branches from the NixOS repo because it's the only sane way for having stacked PRs in GitHub (although that's a bit of an edge-case)

Author: Fwiw this makes no claim about the lifecycle of the feature branch. Any branch that isn't a maintenance branch or - I guess - a proof of concept should be short-lived.
I don't think we should focus any attention on these branches for the time being, while we have bigger fish to fry, such as the again increasing PR queue, and stabilization.

## Reverting

If a change turns out to be merged by mistake, or contain a regression, it may be reverted.

> A revert is not a rejection; just part of an effective development process.

Reviewer: ```suggestion
A revert is not a rejection of the contribution, but merely part of an effective development process.
```

> It makes sure that development keeps running smoothly, with minimal uncertainty, less overhead.

Reviewer: ```suggestion
It makes sure that development keeps running smoothly, with minimal uncertainty, and less overhead.
```

If maintainers have to worry too much about avoiding reverts, they would not be able to merge as much.
By embracing reverts as a good part of the development process, everyone wins.

> However, taking a step back may feel un-fun, so we will try to be extra supportive on the next try.

Reviewer: ```suggestion
However, taking a step back may be frustrating, so maintainers will be extra supportive on the next try.
```
